toml.abnf

```diff
@@ -84,9 +84,10 @@ ml-basic-string = ml-basic-string-delim ml-basic-body ml-basic-string-delim
 
 ml-basic-string-delim = 3quotation-mark
 
-ml-basic-body = *( ml-basic-char / newline / ( escape ws newline ) )
-ml-basic-char = ml-basic-unescaped / escaped
-ml-basic-unescaped = wschar / %x21-5B / %x5D-7E / non-ascii
+ml-basic-body = ws *( ml-basic-char ws / newline / ml-basic-escaped-ws-nl )
+ml-basic-char = ( ml-basic-unescaped / escaped ) ws
```

Comment:
I'm not sure I see how removing whitespace from `ml-basic-unescaped` is trying to do.  It seems to make it a bit more complex, but I don't see why.  Can you say more about what this changes and why?

Also, this seems to duplicate `ws` at the end here, since it is also included in `ml-basic-body`.

Reply:
First off, the README.md is not matching the ABNF:

> When the last non-whitespace character on a line is a \, it will be trimmed along with all whitespace (including newlines) up to the next non-whitespace character or closing delimiter.

So `( escape ws newline )` shoud me something more like `( escape ws newline 1*( wschar / newline ) )`.

So for two successive tokens:
```
( escape ws newline 1*( wschar / newline ) ) ( escape ws newline 1*( wschar / newline ) ) [...]
```

So we can end-up with:

```
( escape ws newline 1*( wschar / newline ) ) ( ml-basic-char ) [...]
```

Or:
```
escape ws newline 1*( wschar / newline ) ( ml-basic-unescaped / escaped ) [...]
```

Or possibly:
```
escape ws newline 1*( wschar / newline ) ml-basic-unescaped [...]
```

Or possibly:
```
escape ws newline 1*wschar ml-basic-unescaped [...]
```

Or:
```
escape ws newline 1*wschar ( wschar / %x21-5B / %x5D-7E / non-ascii ) [...]
```


Or possibly:
```
escape ws newline 1*wschar wschar [...]
```

And then we see that the first ` 1*wschar` is part of  the "escape" token, and the second `wschar` is part of ml-basic-body.

So face to this:

```
escape wschar newline wschar wschar wschar wschar wschar [...]
```

Where do we draw a line between the `wschar` which is part of escape, and the `wschar` which is part of the normal text character (which can also be a space).

Not only there might be this bug in the proposed version, but it might also be present on mine, which is still wrong!

Reply:
I redid that commit to fix it partially and simplify the `ws` mess I had:

```
ml-basic-body = *( ml-basic-char / wschar / newline / ml-basic-escaped-ws-nl )
ml-basic-char = ( ml-basic-unescaped / escaped )
ml-basic-unescaped = %x21-5B / %x5D-7E / non-ascii
ml-basic-nonblank = %x21-7E / non-ascii
ml-basic-escaped-ws-nl = escape newline *( wschar / newline ) ml-basic-nonblank
```

This now matches much more closely the sentence from the `README.md` (yay!):

> When the last non-whitespace character on a line is a `\`, it
> will be trimmed along with all whitespace (including newlines) up to the next
> non-whitespace character **or closing delimiter**.

But it is still wrong as as you can see in bold, this won't work for a \ whitespaces at the end of a basic multiline string.

But to have that in the ABNF definition, we will need to duplicate quite a large proportion, because if we include the closing delimiter in the `ml-basic-escaped-ws-nl`, then we have to *not* include it in the `ml-basic-string` delimiter for not having it duplicated.

So we would need this instead:

```
;; Multiline Basic String

ml-basic-string =  ml-basic-string-delim ml-basic-body ml-basic-string-delim
ml-basic-string =/ ml-basic-string-delim ml-basic-body escape newline *( wschar / newline ) ml-basic-string-delim

ml-basic-string-delim = 3quotation-mark

ml-basic-body = *( ml-basic-char / wschar / newline / ml-basic-escaped-ws-nl )
ml-basic-char = ( ml-basic-unescaped / escaped )
ml-basic-unescaped = %x21-5B / %x5D-7E / non-ascii
ml-basic-escaped-ws-nl = escape newline *( wschar / newline ) ml-basic-nonblank
ml-basic-nonblank = %x21-7E / non-ascii
```